### PR TITLE
feat(ST-42): protect route with auth guard wrapper component

### PR DESCRIPTION
**Pull request by faysal-alam-sazim**
## Description of Change

- Added AuthGuard component inside shared/wrappers
- Used AuthGuard for dashboard page
- Modified redirect implementation logic
- Used useEffect to validate user after login

## Associated Ticket Link(s)

- https://trello.com/c/vteNAEuU

## Related Pull Request(s)

- N/A

## Screenshots / Screen Recordings

https://github.com/user-attachments/assets/8933d803-8a25-43a9-be68-bcccd966d2b2

### CODE REVIEW

File: modules/Login/hooks/useLoginFormData.ts [hunk 1]
```diff
@@ -20,15 +19,16 @@ const useLoginFormData = () => {
...
 
   const onSubmit = async (data: TLoginFormData) => {
     setIsSubmitting(true);
     try {
       const res = await login(data).unwrap();
+
       setInLocalStorage(ACCESS_TOKEN_LOCAL_STORAGE_KEY, res.accessToken);
       dispatch(setUser(res.user));
-      router.push("/dashboard");
+      await getMe().unwrap();
```
Comment (ishmam-iqbal-sazim) [inline modules/Login/hooks/useLoginFormData.ts:31]: Why are we removing the redirection?

File: pages/login/index.page.tsx [hunk 2]
```diff
@@ -1,7 +1,30 @@
...
+    if (user) {
+      if (router.query["redirect"]) {
+        router.push(router.query["redirect"] as string);
+      } else {
+        router.push("/dashboard");
+      }
+    }
+  }, [isLoading, user, router]);
+
+  if (isLoading || user) return <LoadingComponent visible />;
```
Comment (ishmam-iqbal-sazim) [inline pages/login/index.page.tsx:25]: Is this necessary? Successful logic should just route the user to dashboard

File: shared/components/wrappers/AuthGuard/AuthGuard.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    if (!isLoading && isUnauthenticated) {
+      const redirectTo = `${location.pathname}${location.search}`;
+      console.log("Redirect", redirectTo);
+      router.push(getLoginUrlWithRedirectParam(redirectTo));
+    }
+  }, [router, isLoading, error, isUnauthenticated]);
+
+  if (isLoading) return <LoadingComponent visible={true} />;
+
+  if (isUnauthenticated) return <Unauthorized />;
```
Comment (ishmam-iqbal-sazim) [inline shared/components/wrappers/AuthGuard/AuthGuard.tsx:27]: It should just redirect the user to login page. No need to show unauthorized

File: shared/components/wrappers/AuthGuard/AuthGuard.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  const isUnauthenticated = !isLoading && (error || !user);
+
+  useEffect(() => {
+    if (isLoading || typeof location === "undefined") return;
+
+    if (!isLoading && isUnauthenticated) {
+      const redirectTo = `${location.pathname}${location.search}`;
+      console.log("Redirect", redirectTo);
+      router.push(getLoginUrlWithRedirectParam(redirectTo));
+    }
```
Comment (ishmam-iqbal-sazim) [inline shared/components/wrappers/AuthGuard/AuthGuard.tsx:22]: Do we need this redirection logic?

File: shared/components/wrappers/AuthGuard/AuthGuard.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  const router = useRouter();
+  const { isLoading, error, user } = useSessionContext();
+  const isUnauthenticated = !isLoading && (error || !user);
+
+  useEffect(() => {
+    if (isLoading || typeof location === "undefined") return;
+
+    if (!isLoading && isUnauthenticated) {
+      const redirectTo = `${location.pathname}${location.search}`;
+      console.log("Redirect", redirectTo);
```
Comment (ragib-hossain-sazim) [inline shared/components/wrappers/AuthGuard/AuthGuard.tsx:20]: Please remove console logs.

File: pages/login/index.page.tsx (hunk 2 above)
Comment (faysal-alam-sazim) [reply pages/login/index.page.tsx:25]: > Is this necessary? Successful logic should just route the user to dashboard

Not actually necessary as it has only one protected route which is dashboard. Okay I will remove it.

File: modules/Login/hooks/useLoginFormData.ts (hunk 1 above)
Comment (faysal-alam-sazim) [reply modules/Login/hooks/useLoginFormData.ts:31]: > Why are we removing the redirection?

For redirection login. As it is not required right now, will undo it.

**Review by ishmam-iqbal-sazim — APPROVED** [review]

**Review by ragib-hossain-sazim — APPROVED** [review]

**Pull request merged**